Frameworks/CoreGraphics/CGPattern.mm

```diff
@@ -17,160 +17,181 @@
+                CGFloat yStep,
+                CGPatternTiling tiling,
+                bool isColored,
+                const CGPatternCallbacks* callbacks)
+        : _info(info),
+          _bounds(bounds),
+          _transformMatrix(matrix),
+          _xStep(xStep),
+          _yStep(yStep),
+          _tiling(tiling),
+          _isColored(isColored),
+          _callbacks(*callbacks) {
     }
 
-    // TODO: Take into account x/yStep, to justify the two calls here.
-    CGContextRef patternCtx;
-    CGColorSpaceRef colorSpace;
-    CGBitmapInfo bitmapInfo;
-
-    if (isColored) {
-        bitmapInfo = kCGImageAlphaFirst | kCGBitmapByteOrder32Big;
-        colorSpace = CGColorSpaceCreateDeviceRGB();
-    } else {
-        bitmapInfo = kCGImageAlphaNone;
-        colorSpace = CGColorSpaceCreateDeviceGray();
+    ~__CGPattern() {
+        if (_callbacks.releaseInfo) {
+            _callbacks.releaseInfo(_info);
+        }
     }
 
-    patternCtx = CGBitmapContextCreate(nullptr,
-                                       ceilf(width),
-                                       ceilf(height),
-                                       8,
-                                       4 * ceilf(width),
-                                       CGColorSpaceCreateDeviceRGB(),
-                                       kCGImageAlphaPremultipliedFirst | kCGBitmapByteOrder32Big);
-
-    CGContextTranslateCTM(patternCtx, 0.0f, height);
-    CGContextScaleCTM(patternCtx, 1.0f, -1.0f);
-
-    callbacks.drawPattern(info, patternCtx);
-
-    CGImageRef tilePattern = CGBitmapContextCreateImage(patternCtx);
-
-    CGColorSpaceRelease(colorSpace);
-    CGContextRelease(patternCtx);
-
-    float outWidth = width, outHeight = height;
-    // TODO: Min limit on tiling or something?
-    /*      while (outWidth < 256) {
-    outWidth += width;
-    }
-    while (outHeight < 256) {
-    outHeight += height;
-    }*/
-
-    if (isColored) {
-        bitmapInfo = kCGImageAlphaFirst | kCGBitmapByteOrder32Big;
-        colorSpace = CGColorSpaceCreateDeviceRGB();
-    } else {
-        bitmapInfo = kCGImageAlphaNone;
-        colorSpace = CGColorSpaceCreateDeviceGray();
+    inline CGAffineTransform PatternTransform() {
+        return _transformMatrix;
     }
 
-    patternCtx = CGBitmapContextCreate(NULL,
-                                       ceilf(outWidth),
-                                       ceilf(outHeight),
-                                       8,
-                                       4 * ceilf(outWidth),
-                                       CGColorSpaceCreateDeviceRGB(),
-                                       kCGImageAlphaPremultipliedFirst | kCGBitmapByteOrder32Big);
-    CGRect rect = { 0, 0, width, height };
-    CGContextDrawTiledImage(patternCtx, rect, tilePattern);
-
-    CGColorSpaceRelease(colorSpace);
-    CGImageRelease(tilePattern);
-
-    if (generatedImage) {
-        CGImageRelease(generatedImage);
+    inline CGImageRef GetPatternImage() {
```

Comment:
hmm I might as well do it here.

---
In reply to: [108832278](https://github.com/Microsoft/WinObjC/pull/2374#discussion_r108832278) [](ancestors = 108832278)